File: tests/test_builder.cpp
#include "../firmware/components/remote_builder/model.h"
#include <cassert>
#include <iostream>
using namespace airbridge_builder;
int main(){
  Store s{};assert(valid(s));
  std::strcpy(s.remotes[0],"Test remote");auto &b=s.buttons[0];b.active=true;
  assert(!valid(s));std::strcpy(b.name,"Power");assert(valid(s));
  std::vector<int32_t> p(16);for(unsigned i=0;i<p.size();++i)p[i]=(i%2)?-400:400;
  assert(valid_pulses(p));b.count=p.size();for(unsigned i=0;i<p.size();++i)b.pulses[i]=p[i];
  assert(valid(s));assert(unpack(b)==p);
  b.count=257;assert(!valid(s));b.count=16;b.pulses[1]=400;assert(!valid(s));b.pulses[1]=-400;
  b.remote=4;assert(!valid(s));b.remote=0;b.icon=8;assert(!valid(s));b.icon=0;
  b.active=false;assert(!valid(s));b.active=true;s.magic=0;assert(!valid(s));s.magic=MAGIC;
  std::memset(b.name,'x',sizeof(b.name));assert(!valid(s));
  p[0]=-400;assert(!valid_pulses(p));p[0]=400;p[1]=INT32_MIN;assert(!valid_pulses(p));
  assert(!valid_pulses(std::vector<int32_t>(257,400)));
  std::cout<<"Builder storage and pulse validation passed; store bytes="<<sizeof(Store)<<"\n";
}

File: tests/test_builder.js
const assert=require('node:assert/strict');
const {validateRemoteBackup}=require('../firmware/ui/builder.js');
const blank=()=>({schema:2,profile:'433.937MHz ASK/OOK',remotes:['Test','','',''],buttons:Array.from({length:8},(_,slot)=>({slot,name:'',remote:0,icon:0,order:slot,active:false,pulses:[]}))});
let data=blank();validateRemoteBackup(data);
data.buttons[0].active=true;data.buttons[0].name='<img src=x onerror=alert(1)>';
data.buttons[0].pulses=Array.from({length:16},(_,i)=>i%2?-400:400);validateRemoteBackup(data);
for(const mutate of [d=>d.schema=99,d=>d.remotes[0]='',d=>d.buttons[0].slot=7,d=>d.buttons[0].remote=-1,d=>d.buttons[0].icon=8,d=>d.buttons[0].order=8,d=>d.buttons[0].pulses[1]=400,d=>d.buttons[0].pulses[0]=2**32,d=>d.buttons[0].name='\0',d=>d.buttons[0].pulses=Array(257).fill(400),d=>d.buttons[0].name='😀'.repeat(9)]){
  const copy=structuredClone(data);mutate(copy);assert.throws(()=>validateRemoteBackup(copy));
}
console.log('Backup schema, bounds, hostile labels, and pulse validation passed.');
